examples/vision/metric_learning.py

```diff
@@ -22,290 +24,380 @@
 
-    # Double size for visualisation.
-    collage = collage.resize((2 * num_cols * box_size, 2 * num_rows * box_size))
-    return collage
+print(" Create Training Data ".center(34, '#'))
+train_ds = TFDatasetMultiShotMemorySampler(
+    'cifar10',
+    classes_per_batch=classes_per_batch,
+    splits='train',
+    steps_per_epoch=5000,
+    examples_per_class_per_batch=examples_per_class_per_batch,
+    class_list=class_list,
+    preprocess_fn=normalize)
 
-
-# Show a collage of 5x5 random images.
-sample_idxs = np.random.randint(0, 50000, size=(5, 5))
-examples = x_train[sample_idxs]
-show_collage(examples)
+print("\n" + " Create Test Data ".center(34, '#'))
+test_ds = TFDatasetMultiShotMemorySampler('cifar10',
+                                          classes_per_batch=classes_per_batch,
+                                          splits='test',
+                                          total_examples_per_class=100,
+                                          preprocess_fn=normalize)
 
 """
-Metric learning provides training data not as explicit `(X, y)` pairs but instead uses
-multiple instances that are related in the way we want to express similarity. In our
-example we will use instances of the same class to represent similarity; a single
-training instance will not be one image, but a pair of images of the same class. When
-referring to the images in this pair we'll use the common metric learning names of the
-`anchor` (a randomly chosen image) and the `positive` (another randomly chosen image of
-the same class).
-
-To facilitate this we need to build a form of lookup that maps from classes to the
-instances of that class. When generating data for training we will sample from this
-lookup.
+## Visualize the dataset
+
+The samplers will shuffle the dataset, so we can get a sense of the dataset by
+plotting the first 25 images.
+
+The samplers provide a `get_slice(begin, size)` method that allows us to easily
+select a block of samples.
 """
 
-class_idx_to_train_idxs = defaultdict(list)
-for y_train_idx, y in enumerate(y_train):
-    class_idx_to_train_idxs[y].append(y_train_idx)
+num_cols = num_rows = 5
+# Get the first 25 examples.
+X, y = train_ds.get_slice(begin=0, size=num_cols * num_rows)
 
-class_idx_to_test_idxs = defaultdict(list)
-for y_test_idx, y in enumerate(y_test):
-    class_idx_to_test_idxs[y].append(y_test_idx)
+fig = plt.figure(figsize=(6., 6.))
+grid = ImageGrid(fig, 111, nrows_ncols=(num_cols, num_rows), axes_pad=0.1)
+
+for ax, im, label in zip(grid, X, y):
+    ax.imshow(im)
+    ax.axis('off')
 
 """
-For this example we are using the simplest approach to training; a batch will consist of
-`(anchor, positive)` pairs spread across the classes. The goal of learning will be to
-move the anchor and positive pairs closer together and further away from other instances
-in the batch. In this case the batch size will be dictated by the number of classes; for
-CIFAR-10 this is 10.
+## Batch Sizes
```

Comment:
We don't need this -- its a distraction from the colab intent -- instead add a sentence when you talk to get slice.

Reply:
Done